src/main/java/com/commercetools/pspadapter/notification/processor/impl/PaymentSaleNotificationProcessorBase.java

```diff
@@ -35,8 +47,37 @@ public boolean canProcess(@Nonnull Event event) {
                 .equalsIgnoreCase(event.getEventType());
     }
 
-    protected AddTransaction createAddTransactionAction(@Nonnull Event event,
-                                                        @Nonnull TransactionType transactionType) {
+    protected List<? extends UpdateAction<Payment>> createUpdatePaymentActions(@Nonnull Payment ctpPayment, @Nonnull Event event) {
+        String resourceId = getResourceId(event);
+        return findTransactionByInteractionId(ctpPayment.getTransactions(), resourceId)
+                .map(txn -> {
+                    if (!txn.getType().equals(getExpectedTransactionType())) {
+                        logger.error("Found txn paymentId=[{}] with corresponding resourceId={},"
+                                        + " but transactionType=[{}] is not expectedTransactionType=[{}]."
+                                        + " Will create new transaction for the eventId=[{}]",
+                                ctpPayment.getId(), resourceId, getExpectedTransactionType(), event.getId());
+                        return createAddTransactionActionList(event, getExpectedTransactionType());
+                    } else if (isTxnAlreadyUpdated(txn)) {
```

Comment:
I'd put this case
`if (isTxnAlreadyUpdated(txn))`
to `createChangeTransactionStateActionList()`, so the `map` will look simpler.

This looks reasonable that  `createChangeTransactionStateActionList()` completely takes care will it be an empty or filled list.

Actually, i'd say the whole `map` lambda will look better in dedicate named method, because `orElseGet()` block below is lost in all the branched and hard to read

Reply:
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/69/commits/895e4c29c944d028ffa2be3e0c7dd214436357bb